blockchain/pool.go

```diff
@@ -1,630 +1,372 @@
-func (pool *BlockPool) PopRequest() {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
+func (pool blockPool) getMaxPeerHeight() int64 {
+	return pool.maxPeerHeight
+}
 
-	if r := pool.requesters[pool.height]; r != nil {
-		/*  The block can disappear at any time, due to removePeer().
-		if r := pool.requesters[pool.height]; r == nil || r.block == nil {
-			PanicSanity("PopRequest() requires a valid block")
-		}
-		*/
-		r.Stop()
-		delete(pool.requesters, pool.height)
-		pool.height++
-	} else {
-		panic(fmt.Sprintf("Expected requester to pop, got nothing at height %v", pool.height))
-	}
+func (pool *blockPool) reachedMaxHeight() bool {
+	return pool.maxPeerHeight == 0 || pool.height >= pool.maxPeerHeight
 }
 
-// Invalidates the block at pool.height,
-// Remove the peer and redo request from others.
-// Returns the ID of the removed peer.
-func (pool *BlockPool) RedoRequest(height int64) p2p.ID {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	request := pool.requesters[height]
-	peerID := request.getPeerID()
-	if peerID != p2p.ID("") {
-		// RemovePeer will redo all requesters associated with this peer.
-		pool.removePeer(peerID)
-	}
-	return peerID
+func (pool *blockPool) rescheduleRequest(peerID p2p.ID, height int64) {
+	pool.logger.Debug("reschedule requests made to peer for height ", "peerID", peerID, "height", height)
+	pool.requests[height] = true
+	delete(pool.blocks, height)
+	delete(pool.peers[peerID].blocks, height)
 }
 
-// TODO: ensure that blocks come in order for each peer.
-func (pool *BlockPool) AddBlock(peerID p2p.ID, block *types.Block, blockSize int) {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	requester := pool.requesters[block.Height]
-	if requester == nil {
-		pool.Logger.Info("peer sent us a block we didn't expect", "peer", peerID, "curHeight", pool.height, "blockHeight", block.Height)
-		diff := pool.height - block.Height
-		if diff < 0 {
-			diff *= -1
-		}
-		if diff > maxDiffBetweenCurrentAndReceivedBlockHeight {
-			pool.sendError(errors.New("peer sent us a block we didn't expect with a height too far ahead/behind"), peerID)
+// Updates the pool's max height. If no peers are left maxPeerHeight is set to 0.
+func (pool *blockPool) updateMaxPeerHeight() {
+	var newMax int64
+	for _, peer := range pool.peers {
+		if peer.height > newMax {
+			newMax = peer.height
 		}
-		return
 	}
 
-	if requester.setBlock(block, peerID) {
-		atomic.AddInt32(&pool.numPending, -1)
-		peer := pool.peers[peerID]
-		if peer != nil {
-			peer.decrPending(blockSize)
+	if newMax < pool.maxPeerHeight {
+		// Remove any planned requests for heights over the new maxPeerHeight.
+		// This may happen if a peer has updated with lower height.
+		for h := range pool.requests {
+			if h > newMax {
+				delete(pool.requests, h)
+			}
 		}
-	} else {
-		pool.Logger.Info("invalid peer", "peer", peerID, "blockHeight", block.Height)
-		pool.sendError(errors.New("invalid peer"), peerID)
 	}
+	if pool.nextRequestHeight > newMax {
```

Comment:
Here are the steps:
1. All the block requests made to the 150-tall peer or blocks received from the 150-tall peer that are still around (not executed) are removed. Let's say blocks 45..100 (can be a mixture of received and requested)
2. All the heights of these removed blocks are added to the requests slice, i.e. it will be [45..100]
3. The peer is deleted and the new max is set to 50
4. Any of the heights in the requests slice that are higher than the new max are removed from, ending up in this example with [45..50].
5. nextRequestHeight is set to 51.

I have reorganized the code here, hoping it's easier to read..it was all over the place before.
And will add some tests for these next.